packages/dashboard/src/testing/testing-ground/testing-ground.tsx

```diff
@@ -20,6 +111,79 @@ export class TestingGround {
             <iot-dashboard dashboardConfiguration={this.dashboardConfiguration}></iot-dashboard>
           </div>
           <div slot="right">
+            <div>
+              <label htmlFor="selectWidget"> Select widget:</label>
+              <select id="selectWidget" onChange={this.selectWidget}>
+                {this.dashboardConfiguration.widgets.map((widget) => (
+                  <option value={widget.id}>{widget.id.toString() + ', ' + widget.componentTag.toString()}</option>
+                ))}
+              </select>
+            </div>
+            <div>
+              <label>Only list visible annotations</label>
+              <input
+                type="checkbox"
+                checked={this.onlyListVisibleAnnotations}
+                onChange={this.onListVisibleAnnotations}
+              />
+            </div>
+            <br />
+            <br />
+            <div>
+              <form onSubmit={this.addAnnotation}>
+                <label htmlFor="addAnnotation"> Add annotation -</label>
+                <label htmlFor="addLabel"> Label:</label>
+                <input type="text" id="addLabel"></input>
+                <input type="color" value="#FF0000"></input>
+                <input type="datetime-local" step="1"></input>
+                <input type="submit" value="Add" id="addAnnotation" onSubmit={this.addAnnotation}></input>
+              </form>
+            </div>
+            <br />
+            <br />
+            <div>
+              <form onSubmit={this.deleteAnnotation}>
+                <label htmlFor="deleteAnnotation"> Delete annotation:</label>
+                <select id="deleteAnnotation">
+                  {this.onlyListVisibleAnnotations
+                    ? this.dashboardConfiguration.widgets
+                        .find((x) => x.id == this.selectedWidgetId)
+                        ?.annotations?.x?.filter((annotation) => this.isVisible(annotation))
+                        .map((annotation) => <option value={annotation.id}>{annotation.value.toString()}</option>)
+                    : this.dashboardConfiguration.widgets
+                        .find((x) => x.id == this.selectedWidgetId)
+                        ?.annotations?.x?.map((annotation) => (
+                          <option value={annotation.id}>{annotation.value.toString()}</option>
+                        ))}
+                </select>
+                <input type="submit" value="Delete"></input>
+              </form>
+            </div>
+            <br />
+            <br />
+            <div>
+              <form onSubmit={this.editAnnotation}>
+                <label htmlFor="editAnnotation"> Edit annotation - </label>
+                <label htmlFor="addLabel"> Label:</label>
+                <input type="text" id="addLabel"></input>
+                <input type="color" value="#FF0000"></input>
+                <br />
+                <select id="editAnnotation">
+                  {this.onlyListVisibleAnnotations
```

Comment:
Same here. We are doing a lot of stuff here that's duplicative of the above ternary as well, so better to abstract these out above the JSX.

Reply:
Refactored; moved filter and find logic above the render function